src/readAndTokenize.h

```diff
@@ -0,0 +1,52 @@
+#ifndef READANDTOKENIZE_H
+#define READANDTOKENIZE_H
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <range/v3/all.hpp>
+#include "tokenizer.h"
+#include "readFile.h"
+
+namespace readAndTokenize {
+  auto validateArgs = [](int argc, char* argv[]) -> std::optional<std::vector<std::string>> {
+    // es müssen genau drei Dateien übergeben werden (book, war terms, peace terms)
+    if (argc != 4) {
+        return std::nullopt;
+    }
+    std::vector<std::string> args(argv, argv + argc);
+    return args;
+  };
+
+  auto readAndTokenize = [](const std::string path) -> std::optional<std::vector<std::string>> {
+      auto file = readFile::readFile(path);
+      if (!file) {
+          return std::nullopt;
+      }
+      return tokenizer::tokenize(*file);
+  };
+
+  auto validateAndTokenize = [](int argc, char* argv[]) -> std::optional<std::vector<std::vector<std::string>>>{
+    auto args = validateArgs(argc, argv);
+    if (!args.has_value()) {
+      return std::nullopt;
+    }
+    std::vector<std::vector<std::string>> tokens = args.value() | ranges::views::drop(1)
+      | ranges::views::transform([](const std::string& arg) {
+          auto token = readAndTokenize(arg);
+          return token.has_value() ? token.value() : std::vector<std::string>{};
+      })
+      | ranges::views::take_while([](const std::vector<std::string>& token) {
+          return !token.empty();
+      })
+      | ranges::to<std::vector<std::vector<std::string>>>;
+
+    if (tokens.size() < (args.value().size() - 1)) {
+      return std::nullopt;
+    }
+
+    return tokens;
+  };
+}
```

Comment:
Yes ranges :)

Es gibt aber bisschen Optimierungspotenzial:

Zuerst versuchst du ja die Tokens rauszulesen und speicherst dann entweder die value oder legst dann nochmal einen leeren Vektor an. Da du die angelegten Vektoren im nächsten Schritt aber verwirfst ist das bisschen unnötiger Overhead. 

Du kannst z.B. erst die token**s** auslesen, dann die Ergebnisse filtern, ob es denn überhaupt ein Ergebnis gibt und dann erst in einem Vektor zusammenführen aka:

```c++
std::vector<std::vector<std::string>> tokens = args.value() | ranges::views::drop(1)
    | ranges::views::transform([](const std::string& arg) -> std::optional<std::vector<std::string>> {
        return readAndTokenize(arg);
    })
    | ranges::views::filter([](const std::optional<std::vector<std::string>>& optTokens) {
        return optTokens.has_value();
    })
    | ranges::views::transform([](const std::optional<std::vector<std::string>>& tokens) {
        return tokens.value();
    })
    | ranges::to<std::vector<std::vector<std::string>>>;
```

Das erste Argument würde ich übrigens auch noch zwecks Separation of concerns schon in der validateArgs Methode rausschmeißen :)